R-packages/evalcast/R/get_predictions.R

```diff
@@ -69,22 +81,36 @@ get_predictions <- function(forecaster,
                             apply_corrections = function(signals) signals,
                             response_data_source = signals$data_source[1],
                             response_data_signal = signals$signal[1],
-                            forecaster_args = list()) {
-  
+                            parallel_execution = TRUE,
+                            forecaster_args = list(),
+                            honest_as_of = TRUE,
+                            offline_signal_dir = NULL) {
+
   assert_that(is_tibble(signals), msg="`signals` should be a tibble.")
-  
-  
-  out <- forecast_dates %>%
-    map(~ do.call(
-          get_predictions_single_date,
-          list(forecaster = forecaster,
-               signals = signals,
-               forecast_date = .x,
-               apply_corrections = apply_corrections,
-               forecaster_args = forecaster_args)
-          )
-        ) %>%
-    bind_rows()
+  assert_that(xor(honest_as_of, "as_of" %in% names(signals)), msg="`honest_as_of` should be set if and only if `as_of` is not set. Either remove as_of specification or set honest_as_of to FALSE.")
+
+  p = progressor(steps = length(forecast_dates))
+  get_predictions_single_date_ <- function(forecast_date) {
+    p()
+    do.call(get_predictions_single_date,
+            list(forecaster = forecaster,
+              signals = signals,
+              forecast_date = forecast_date,
+              apply_corrections = apply_corrections,
+              forecaster_args = forecaster_args,
+              honest_as_of = honest_as_of,
+              offline_signal_dir = offline_signal_dir))
+  }
+  if(parallel_execution) {
+    out <- mclapply(
+      forecast_dates, 
+      get_predictions_single_date_, 
+      mc.cores=max(detectCores()-1, 1)
```

Comment:
`mc.cores=max(detectCores()-1, 1)` should be something like
- `getOption("mc.cores", 2L)` --- the default value of this argument
- `getOption("mc.cores", max(detectCores()-1, 1))` --- uses the current value unless the user overrides beforehand using `options(mc.cores=<value>)`
- changing `get_forecasts` to take an integer number of cores rather than a Boolean whether to use parallelism

Reply:
In favor of the second one, so the user doesn't have to figure out their core number by default, but can also set `mc.cores` manually.

Reply:
The third option could also be modified to have some user input value mean "auto-select some number of cores for me greater than 1 if possible".  Then we'd avoid the need to go through the side-effectful `options` interface while allowing the user to not have to figure out the core number.  But this is just trading one confusing/complicated thing for another, unless you can think of a good name and scheme for such a parameter for option 3'.  If you can think of a good configuration for 3', that might be preferable.

If proceeding with option 2, it'd be good to mention the possibility of using `options(mc.cores=<...>)` (or `withr::with_options(......)`) in the function's documentation.

Reply:
What if `parallel_execute` is a mixed Boolean+Integer type, that if TRUE defaults to `detectCores()-1` and if an integer then uses that many cores?

Reply:
Implemented the above idea in the latest commit.